Applying the strategy 10 to bFactorList seems to decode the list of

JulienFerte opened the pull request:
floating numbers, although one needs to check whether the values are correct. On the other hand, decode_strategy_into_float_stuff returning the pointer value makes a segfault on my machine. Does it crash on the other's machine? I need to check things out but I have no clues on what is going on.

Commit:
Corrected an error message.

Commit:
Applying the strategy 10 to bFactorList seems to decode the list of floating numbers, although one needs to check whether the values are correct. On the other hand, returning the value makes a segfault on my machine. Does it crash on the other's machine? If so, I need to check things out but I have no clues on what is going on.

Commit:
Clearing a bit the code. What is actually segfaulting on my machine is not the decode_strategy_stuff, it is the free(buffer) from msgpack. So nothing new there.

abradle commented:
Hi Juliean,

Great - thank you. 

Could I ask that you do `git add -u` rather than `git add *` and stop adding in all the Cmake files. This is 1) because it stops the continuous integration from working and 2) becaues it makes it hard to see what you've commited.

Best wishes,

Anthony

JulienFerte commented:
Noted. I am working on the translation from an MMTF file into the C 
struct.

Before jumping into writing this long piece of code, I would like to 
have things cristal clear.

Could you point me to the Java file implementing the Java version of 
this translation? (Maybe Alexandr knows which file it is.)

I was wondering: depending on the MMTF file, the strategy 
encoding/decoding a given field can vary?

Schematically, a field F has the type T (float_, char_, etc), then the 
strategy S is given by the MMTF file and the decoding is as follows:
byte\* ----[applying strategy S, with parameter if present]---> type T

Am I correct?

I will organize the decoding into functions: mmtf_sample.[field] = 
fetch_field_[output_type](char* name, params).

On 18.07.2016 16:39, Anthony Bradley wrote:

> Hi Juliean,
> 
> Great - thank you.
> 
> Could I ask that you do git add -u rather than git add \* and stop
> adding in all the Cmake files. This is 1) because it stops the
> continuous integration from working and 2) becaues it makes it hard to
> see what you've commited.
> 
> Best wishes,
> 
> Anthony
> 
> ## 
> 
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or mute the
> thread [2].
> 
>   *
> 
> ## Links:
> 
> [1] https://github.com/rcsb/mmtf-c/pull/8#issuecomment-233348067
> [2]
> https://github.com/notifications/unsubscribe-auth/ATa62AIHxHkQdM2wgqE5q8VDPP9PfzARks5qW5ARgaJpZM4JOmJd

## 

Julien Ferté

http://www.julienferte.com/

Commit:
Fixed the wrong floating point numbers decoded. Now the bFactorList decoded matches with the values in the file 4HHB.pdb

Although there is still a trouble of floating-point rounding. This needs to be fixed.

Added the functions fetch_field_[output_type]( msgpack object, field name )

JulienFerte commented:
Hi,

I fixed the wrong floating point numbers decoded. Now the bFactorList 
decoded matches with the values in the file 4HHB.pdb. Although there is 
still a trouble of floating-point rounding. This needs to be fixed.

Added the functions fetch_field_[output_type](msgpack object, field 
name)

The C library is about to be done.

On 18.07.2016 16:39, Anthony Bradley wrote:

> Hi Juliean,
> 
> Great - thank you.
> 
> Could I ask that you do git add -u rather than git add \* and stop
> adding in all the Cmake files. This is 1) because it stops the
> continuous integration from working and 2) becaues it makes it hard to
> see what you've commited.
> 
> Best wishes,
> 
> Anthony
> 
> ## 
> 
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or mute the
> thread [2].
> 
>   *
> 
> ## Links:
> 
> [1] https://github.com/rcsb/mmtf-c/pull/8#issuecomment-233348067
> [2]
> https://github.com/notifications/unsubscribe-auth/ATa62AIHxHkQdM2wgqE5q8VDPP9PfzARks5qW5ARgaJpZM4JOmJd

## 

Julien Ferté

http://www.julienferte.com/

abradle commented:
This is really awesome Julien.

https://github.com/rcsb/mmtf-java/blob/master/mmtf-codec/src/main/java/org/rcsb/mmtf/decoder/GenericDecoder.java

Is the file you're looking for.

JulienFerte commented:
Thanks for the link.

After a few coding and testing, the function get_len you wrote only gets 
the length of the decoded array. Not of the encoded array.

The length of the encoded array can be retrieved from the MsgPack 
headers. I am currently modifying the code in that sense.

On 19.07.2016 16:36, Anthony Bradley wrote:

> This is really awesome Julien.
> 
> https://github.com/rcsb/mmtf-java/blob/master/mmtf-codec/src/main/java/org/rcsb/mmtf/decoder/GenericDecoder.java
> [1]
> 
> Is the file you're looking for.
> 
> ## 
> 
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [2], or mute the
> thread [3].
> 
>   *
> 
> ## Links:
> 
> [1]
> https://github.com/rcsb/mmtf-java/blob/master/mmtf-codec/src/main/java/org/rcsb/mmtf/decoder/GenericDecoder.java
> [2] https://github.com/rcsb/mmtf-c/pull/8#issuecomment-233652949
> [3]
> https://github.com/notifications/unsubscribe-auth/ATa62OPvOwVhP9wqzcKENRtFVzVJCJSUks5qXOD5gaJpZM4JOmJd

## 

Julien Ferté

http://www.julienferte.com/

JulienFerte commented:
Hi.

By using the proper msgpack length, I have got all the accurate floats 
from x,y,zCoords and bFactors... and more.

There are actually 6 extra floating point numbers for each computed 
list. Is it normal?

For example, after decoding, the xCoordList ends with:

[...]
4778 : -1.263000
4779 : -23.158001
4780 : -5.895000
4781 : 22.635000
4782 : 48.311001
4783 : 75.306000
4784 : 105.198997

In the PDB file, the value -1.263 is the xCoordinate of the last atom. 
What are the 6 extra floats?

The numbers 4778 and so on denotes the rank in the C array containing 
the decoded values.
Counting the rank in a C array starts from 0, and the MMTF file arrays 
don't contain any slot for the TER entries from the PDB, hence the 
difference in ranking between the PDB ranking of an atom and the ranking 
given by the strategy decoders.

What is also surprising is that the get_len function gets the value 4779 
from the header of the byte array contained within the msgpack binary 
data (from the bytes 4 to 7). Since the C array is numbered starting 
from 0: rank 4778 means 4779 in length. This value 4779 is accurate.

On 19.07.2016 16:36, Anthony Bradley wrote:

> This is really awesome Julien.
> 
> https://github.com/rcsb/mmtf-java/blob/master/mmtf-codec/src/main/java/org/rcsb/mmtf/decoder/GenericDecoder.java
> [1]
> 
> Is the file you're looking for.
> 
> ## 
> 
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [2], or mute the
> thread [3].
> 
>   *
> 
> ## Links:
> 
> [1]
> https://github.com/rcsb/mmtf-java/blob/master/mmtf-codec/src/main/java/org/rcsb/mmtf/decoder/GenericDecoder.java
> [2] https://github.com/rcsb/mmtf-c/pull/8#issuecomment-233652949
> [3]
> https://github.com/notifications/unsubscribe-auth/ATa62OPvOwVhP9wqzcKENRtFVzVJCJSUks5qXOD5gaJpZM4JOmJd

## 

Julien Ferté

http://www.julienferte.com/

abradle commented:
Hi Julien,

This is excellent.

So yes get_len gets the length of the decoded array.

I think the reason you're getting 6 extra values in your decoded array is because you are not considering the 12 byte header at the start of the array. (This is a header we have written). I suspect if for each array you use the messagepack length -12 as the length of the encoded array - you will get the correct answer.

I hope this makes sense! 

Anthony

Commit:
Solve the bloody permanent SEGFAULT problem

JulienFerte commented:
Hi,

I solved the SEGFAULT problem which had been occurring on my machine.

I was wondering two things:
- some MMTF fields are string, ie char arrays, and to decode them from 
  MsgPack, only a cast from a byte into a char is needed?
- how do I decode entityList and groupList? In C they both are of a 
  different type from the Java structure found here:

https://github.com/rcsb/mmtf-java/blob/master/mmtf-codec/src/main/java/org/rcsb/mmtf/decoder/GenericDecoder.java

On 22.07.2016 17:35, Anthony Bradley wrote:

> Hi Julien,
> 
> This is excellent.
> 
> So yes get_len gets the length of the decoded array.
> 
> I think the reason you're getting 6 extra values in your decoded array
> is because you are not considering the 12 byte header at the start of
> the array. (This is a header we have written). I suspect if for each
> array you use the messagepack length -12 as the length of the encoded
> array - you will get the correct answer.
> 
> I hope this makes sense!
> 
> Anthony
> 
> ## 
> 
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or mute the
> thread [2].
> 
>   *
> 
> ## Links:
> 
> [1] https://github.com/rcsb/mmtf-c/pull/8#issuecomment-234576966
> [2]
> https://github.com/notifications/unsubscribe-auth/ATa62NQ3aQLTwHxee_i_Yl2MRA1VLxhlks5qYOM0gaJpZM4JOmJd

## 

Julien Ferté

http://www.julienferte.com/

JulienFerte commented:
Hi,

up for the questions.

I am sorry it took a bit of time but I was dealing with my main project 
MutaFrame, here in Belgium.

I just made a pull request even though there are quite some rubbish 
files which I used during debugging. I will clean everything once the 
decoder is finished.

By the way are you on vacation?

I will be on vacation from the 8th until the 21st August. So, during 
that time, I will most probably pause the work on the decoder. I want to 
finish the lib and I will resume the work after my vacation.

Is it OK for you?

Bye,
Julien

-------- Original Message --------
Subject: Re: [rcsb/mmtf-c] Applying the strategy 10 to bFactorList seems 
to decode the list of (#8)
Date: 28.07.2016 12:47
 From: jferte Julien.Ferte@ulb.ac.be
To: rcsb/mmtf-c 
reply@reply.github.com

Hi,

I solved the SEGFAULT problem which had been occurring on my machine.

I was wondering two things:
- some MMTF fields are string, ie char arrays, and to decode them from 
  MsgPack, only a cast from a byte into a char is needed?
- how do I decode entityList and groupList? In C they both are of a 
  different type from the Java structure found here:

https://github.com/rcsb/mmtf-java/blob/master/mmtf-codec/src/main/java/org/rcsb/mmtf/decoder/GenericDecoder.java

On 22.07.2016 17:35, Anthony Bradley wrote:

> Hi Julien,
> 
> This is excellent.
> 
> So yes get_len gets the length of the decoded array.
> 
> I think the reason you're getting 6 extra values in your decoded array
> is because you are not considering the 12 byte header at the start of
> the array. (This is a header we have written). I suspect if for each
> array you use the messagepack length -12 as the length of the encoded
> array - you will get the correct answer.
> 
> I hope this makes sense!
> 
> Anthony
> 
> ## 
> 
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or mute the
> thread [2].
> 
>   *
> 
> ## Links:
> 
> [1] https://github.com/rcsb/mmtf-c/pull/8#issuecomment-234576966
> [2]
> https://github.com/notifications/unsubscribe-auth/ATa62NQ3aQLTwHxee_i_Yl2MRA1VLxhlks5qYOM0gaJpZM4JOmJd

## 

Julien Ferté

http://www.julienferte.com/

abradle commented:
Hi Julien,

Great stuff!

Timeline sounds very good! Have a lovely holiday!
- some MMTF fields are string, ie char arrays, and to decode them from
  MsgPack, only a cast from a byte into a char is needed?

Yes that sounds good!
- how do I decode entityList and groupList? In C they both are of a
  different type from the Java structure found here:

So entityList and groupList are lists of objects - In Java they are defined here (https://github.com/rcsb/mmtf-java/blob/master/mmtf-serialization/src/main/java/org/rcsb/mmtf/dataholders/Entity.java) and (https://github.com/rcsb/mmtf-java/blob/master/mmtf-serialization/src/main/java/org/rcsb/mmtf/dataholders/Group.java).

A struct that looks a bit like that would be awesome - but if this is not possible we can discuss other options.

Thanks so much for all your work on this! You're doing a great job!

abradle commented:
Sorry - accidently closed

JulienFerte commented:
Hi,

the structures for GroupList and Entity were already in the file 
mmtf_structure.c. After the vacation, my head is more fresh and I 
gathered the answer to my question.

Although I am working on a tedious part of my main project, I am 
recalling step by step the MMTF stuff and will be working on it by the 
end of the week.

## 

Julien

On 02.08.2016 19:14, Anthony Bradley wrote:

> Hi Julien,
> 
> Great stuff!
> 
> Timeline sounds very good! Have a lovely holiday!
> - some MMTF fields are string, ie char arrays, and to decode them
>   from MsgPack, only a cast from a byte into a char is needed?
> 
> Yes that sounds good!
> - how do I decode entityList and groupList? In C they both are of a
>   different type from the Java structure found here:
> 
> So entityList and groupList are lists of objects - In Java they are
> defined here
> (https://github.com/rcsb/mmtf-java/blob/master/mmtf-serialization/src/main/java/org/rcsb/mmtf/dataholders/Entity.java
> [1]) and
> (https://github.com/rcsb/mmtf-java/blob/master/mmtf-serialization/src/main/java/org/rcsb/mmtf/dataholders/Group.java
> [2]).
> 
> A struct that looks a bit like that would be awesome - but if this is
> not possible we can discuss other options.
> 
> Thanks so much for all your work on this! You're doing a great job!
> 
> ## 
> 
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [3], or mute the
> thread [4].
> 
>   *
> 
> ## Links:
> 
> [1]
> https://github.com/rcsb/mmtf-java/blob/master/mmtf-serialization/src/main/java/org/rcsb/mmtf/dataholders/Entity.java
> [2]
> https://github.com/rcsb/mmtf-java/blob/master/mmtf-serialization/src/main/java/org/rcsb/mmtf/dataholders/Group.java
> [3] https://github.com/rcsb/mmtf-c/pull/8#issuecomment-236974235
> [4]
> https://github.com/notifications/unsubscribe-auth/ATa62OqW8rHo5NIEmlU6wpuEq0LYatcaks5qb3rZgaJpZM4JOmJd

## 

Julien Ferté

http://www.julienferte.com/

abradle commented:
Awesome! Thanks Julien! Hope you had good holidays!

Commit:
Created the folder src_v2 where the code is reorganized and cleaned.
I don't know if it compiles through cmake and make, I know it compiles perfectly with my small bash script "./compile.sh".
I managed to decode all fields except for 8 of them.
Those will take quite some time and reorganisation of the code to make.

JulienFerte commented:
Hi,

I have been able to decypher nearly all fields but 8 of them, from the 
MMTF structure.

The last ones are challenging and I will need quite some time and code 
reorganization to programm the last functions.

I have made a folder src_v2/ to reorder and clean the code. I also 
deleted some functions which I don't need, like in the num_converters, 
this file is actually deleted. This is why I made a duplicate of the 
folder src/. Do someone else need these functions?

src_v2 comes along with a simple bash script ./compile_and_run.sh which 
compiles and runs the executable. Inside src_v2 there is also the MMTF 
sample file to decode. The bash script needs msgpack installed on the 
machine as a library for the C compiler gcc, the msgpackc library is 
available on Linux, I think it is available on OSX and Windows also. 
Thus src_v2 doesn't need a copy of the msgpack sources for compiling. It 
is one step closer to the final MMTF decoder.

I don't know if cmake and make are needed for compiling on the other 
commiters' machines, but src_v2 is autonomous and I will go on with 
src_v2/.

Tell me if executing the bash script ./compile_and_run.sh, when you are 
in the folder src_v2/, works on you machine.

I have never programmed with cmake nor make before and, sorry, but one 
thing at a time, decyphering the cryptical documentation on msgpack is 
already nightmare-ish. And I don't currently need cmake nor make, the 
bash script is enough for me.

So, do you need to keep src_v2 synchronized with cmake and make?

On 23.08.2016 22:43, Anthony Bradley wrote:

> Awesome! Thanks Julien! Hope you had good holidays!
> 
> ## 
> 
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or mute the
> thread [2].
> 
>   *
> 
> ## Links:
> 
> [1] https://github.com/rcsb/mmtf-c/pull/8#issuecomment-241872203
> [2]
> https://github.com/notifications/unsubscribe-auth/ATa62KzjpMzrxDxfKnkk0WOmcvi0VPM2ks5qi1t7gaJpZM4JOmJd

## 

Julien Ferté

http://www.julienferte.com/

Commit:
The version 1.0 of the MMTF parser is done.

abradle commented:
Awesome Julien! This looks very good. I will spend a few days working through it. 

And then:
1. Clean up a few of the folders
2. See if I can get cmake to work

Thank you so much for your efforts on this! I will be back in touch if I have questions.
